src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -0,0 +1,574 @@
+    def doErase(tp: Type): Type = tp match {
+      case tp: RefinedType => erase(tp.parent)
+      case _ => tp
+    }
+
+    val origin = doErase(tp)
+    if (origin =:= defn.ArrayType) tp else origin
+  }
+
+  /** Is `tp1` a subtype of `tp2`?  */
+  def isSubType(tp1: Type, tp2: Type): Boolean = {
+    // expose is important due to type bounds
+    // check SI-9657 and tests/patmat/gadt.scala
+    tp1 <:< expose(tp2)
+  }
+
+  def isEqualType(tp1: Type, tp2: Type): Boolean = tp1 =:= tp2
+
+  /** Parameter types of the case class type `tp`  */
+  def signature(tp: Type): List[Type] = {
+    val ktor = tp.classSymbol.primaryConstructor.info
+
+    val meth =
+      if (ktor.isInstanceOf[MethodType]) ktor
+      else
+      tp match {
+        case AppliedType(_, params) =>
+          val refined = params.map {
+            // TypeBounds would generate an exception
+            case tp: TypeBounds => tp.underlying
+            case tp => tp
+          }
+          ktor.appliedTo(refined)
+        case _ =>
+          ktor
+      }
+
+    // refine path-dependent type in params. refer to t9672
+    meth.firstParamTypes.map(_.stripTypeVar).map { ptp =>
+      (tp, ptp) match {
+        case (TypeRef(ref1: TypeProxy, _), TypeRef(ref2: TypeProxy, name)) =>
+          if (ref1.underlying <:< ref2.underlying) TypeRef(ref1, name) else ptp
+        case _ => ptp
+      }
+    }
+  }
+
+  def partitions(tp: Type): List[Space] = {
+    val children = tp.classSymbol.annotations.filter(_.symbol == ctx.definitions.ChildAnnot).map { annot =>
+      // refer to definition of Annotation.makeChild
+      annot.tree match {
+        case Apply(TypeApply(_, List(tpTree)), _) => tpTree.symbol
+      }
+    }
+
+    tp match {
+      case OrType(tp1, tp2) => List(Typ(tp1, true), Typ(tp2, true))
+      case _ if tp =:= ctx.definitions.BooleanType =>
+        List(
+          Const(Constant(true), ctx.definitions.BooleanType),
+          Const(Constant(false), ctx.definitions.BooleanType)
+        )
+      case _ if tp.classSymbol.is(Enum) =>
+        children.map(sym => Const(Constant(sym), tp))
+      case _ =>
+        val parts = children.map { sym =>
+          if (sym.is(ModuleClass))
+            sym.asClass.classInfo.selfType
+          else if (sym.info.typeParams.length > 0 || tp.isInstanceOf[TypeRef])
+            refine(tp, sym.asClass.classInfo.symbolicTypeRef)
+          else
+            sym.typeRef
+        } filter { tpe =>
+          // Child class may not always be subtype of parent:
+          // GADT & path-dependent types
+          isSubType(tpe, tp)
+        }
+
+        parts.map(Typ(_, true))
+    }
+  }
+
+  /** Refine tp2 based on tp1
+   *
+   *  E.g. if `tp1` is `Option[Int]`, `tp2` is `Some`, then return
+   *  `Some[Int]`.
+   *
+   *  If `tp1` is `path1.A`, `tp2` is `path2.B`, and `path1` is subtype of
+   *  `path2`, then return `path1.B`.
+   */
+  def refine(tp1: Type, tp2: Type): Type = (tp1, tp2) match {
```

Comment:
yeah `asSeenFrom` does not make sense here I meant to say something like `(tp1 & tp2).baseTypeWithArgs(tp2.classSymbol)`, no idea if that would work here.